File: animated-base/src/test/java/com/facebook/fresco/animation/bitmap/cache/FrescoFrameCacheTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.fresco.animation.bitmap.cache;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

import android.graphics.Bitmap;
import com.facebook.common.references.CloseableReference;
import com.facebook.imagepipeline.image.CloseableImage;
import com.facebook.imagepipeline.image.CloseableStaticBitmap;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

/**
 * Tests {@link FrescoFrameCache}.
 */
@RunWith(PowerMockRunner.class)
@PrepareForTest(CloseableReference.class)
public class FrescoFrameCacheTest {

  @Mock public CloseableReference<CloseableImage> mImageReference;
  @Mock public CloseableStaticBitmap mCloseableStaticBitmap;
  @Mock public CloseableReference<Bitmap> mBitmapReference;
  @Mock public CloseableReference<Bitmap> mBitmapReferenceClone;
  @Mock public Bitmap mUnderlyingBitmap;

  @Before
  public void setup() {
    MockitoAnnotations.initMocks(this);
    when(mBitmapReference.isValid()).thenReturn(true);
    when(mBitmapReference.get()).thenReturn(mUnderlyingBitmap);

    when(mBitmapReferenceClone.isValid()).thenReturn(true);
    when(mBitmapReferenceClone.get()).thenReturn(mUnderlyingBitmap);

    when(mCloseableStaticBitmap.isClosed()).thenReturn(false);
    when(mCloseableStaticBitmap.getUnderlyingBitmap()).thenReturn(mUnderlyingBitmap);
    when(mCloseableStaticBitmap.convertToBitmapReference())
        .thenReturn(mBitmapReference);
    when(mCloseableStaticBitmap.cloneUnderlyingBitmapReference()).thenReturn(mBitmapReferenceClone);

    when(mImageReference.isValid()).thenReturn(true);
    when(mImageReference.get()).thenReturn(mCloseableStaticBitmap);
  }

  @Test
  public void testExtractAndClose() throws Exception {
    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(mImageReference);

    assertThat(extractedReference).isNotNull();
    assertThat(extractedReference.get()).isEqualTo(mUnderlyingBitmap);
    verify(mImageReference).close();

    extractedReference.close();
  }

  @Test
  public void testExtractAndClose_whenBitmapRecycled_thenReturnReference() throws Exception {
    when(mUnderlyingBitmap.isRecycled()).thenReturn(true);

    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(mImageReference);

    // We only detach the reference and do not care if the bitmap is valid
    assertThat(extractedReference).isNotNull();
    assertThat(extractedReference.get()).isEqualTo(mUnderlyingBitmap);
    verify(mImageReference).close();

    extractedReference.close();
  }

  @Test
  public void testExtractAndClose_whenBitmapReferenceInvalid_thenReturnReference()
      throws Exception {
    when(mBitmapReference.isValid()).thenReturn(false);

    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(mImageReference);

    // We only detach the reference and do not care if the bitmap reference is valid
    assertThat(extractedReference).isNotNull();
    assertThat(extractedReference.get()).isEqualTo(mUnderlyingBitmap);

    extractedReference.close();

    verify(mImageReference).close();
  }

  @Test
  public void testExtractAndClose_whenCloseableStaticBitmapClosed_thenReturnNull()
      throws Exception {
    when(mCloseableStaticBitmap.isClosed()).thenReturn(true);
    when(mCloseableStaticBitmap.cloneUnderlyingBitmapReference()).thenReturn(null);

    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(mImageReference);

    // We only detach the reference and do not care if the bitmap is valid
    assertThat(extractedReference).isNull();
    verify(mImageReference).close();
  }

  @Test
  public void testExtractAndClose_whenImageReferenceInvalid_thenReturnNull() throws Exception {
    when(mImageReference.isValid()).thenReturn(false);

    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(mImageReference);

    // We only detach the reference and do not care if the bitmap is valid
    assertThat(extractedReference).isNull();
    verify(mImageReference).close();
  }

  @Test
  public void testExtractAndClose_whenInputNull_thenReturnNull() throws Exception {
    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(null);

    assertThat(extractedReference).isNull();
    verifyZeroInteractions(mImageReference);
  }

  @Test
  public void testExtractAndClose_whenCloseableStaticBitmapNull_thenReturnNull() throws Exception {
    when(mImageReference.get()).thenReturn(null);

    CloseableReference<Bitmap> extractedReference =
        FrescoFrameCache.convertToBitmapReferenceAndClose(mImageReference);

    assertThat(extractedReference).isNull();
    verify(mImageReference).close();
  }
}

File: animated-base/src/test/java/com/facebook/fresco/animation/bitmap/cache/FrescoFrameCacheTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.animation.bitmap.cache

import android.graphics.Bitmap
import com.facebook.common.references.CloseableReference
import com.facebook.imagepipeline.image.CloseableImage
import com.facebook.imagepipeline.image.CloseableStaticBitmap
import org.assertj.core.api.Assertions
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner

/** Tests [FrescoFrameCache]. */
@RunWith(RobolectricTestRunner::class)
class FrescoFrameCacheTest {
  private lateinit var imageReference: CloseableReference<CloseableImage>
  private lateinit var closeableStaticBitmap: CloseableStaticBitmap
  private lateinit var bitmapReference: CloseableReference<Bitmap>
  private lateinit var bitmapReferenceClone: CloseableReference<Bitmap>
  private lateinit var underlyingBitmap: Bitmap

  @Before
  fun setup() {
    imageReference = mock()
    closeableStaticBitmap = mock()
    bitmapReference = mock()
    bitmapReferenceClone = mock()
    underlyingBitmap = mock()

    whenever(bitmapReference.isValid).thenReturn(true)
    whenever(bitmapReference.get()).thenReturn(underlyingBitmap)

    whenever(bitmapReferenceClone.isValid).thenReturn(true)
    whenever(bitmapReferenceClone.get()).thenReturn(underlyingBitmap)

    whenever(closeableStaticBitmap.isClosed()).thenReturn(false)
    whenever(closeableStaticBitmap.getUnderlyingBitmap()).thenReturn(underlyingBitmap)
    whenever(closeableStaticBitmap.convertToBitmapReference()).thenReturn(bitmapReference)
    whenever(closeableStaticBitmap.cloneUnderlyingBitmapReference())
        .thenReturn(bitmapReferenceClone)

    whenever(imageReference.isValid).thenReturn(true)
    whenever(imageReference.get()).thenReturn(closeableStaticBitmap)
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose() {
    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(imageReference)

    Assertions.assertThat(extractedReference).isNotNull()
    extractedReference?.let { ref ->
      Assertions.assertThat(ref.get()).isEqualTo(underlyingBitmap)
      ref.close()
    }
    verify(imageReference).close()
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose_whenBitmapRecycled_thenReturnReference() {
    whenever(underlyingBitmap.isRecycled).thenReturn(true)

    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(imageReference)

    // We only detach the reference and do not care if the bitmap is valid
    Assertions.assertThat(extractedReference).isNotNull()
    extractedReference?.let { ref ->
      Assertions.assertThat(ref.get()).isEqualTo(underlyingBitmap)
      ref.close()
    }
    verify(imageReference).close()
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose_whenBitmapReferenceInvalid_thenReturnReference() {
    whenever(bitmapReference.isValid).thenReturn(false)

    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(imageReference)

    // We only detach the reference and do not care if the bitmap reference is valid
    Assertions.assertThat(extractedReference).isNotNull()
    extractedReference?.let { ref ->
      Assertions.assertThat(ref.get()).isEqualTo(underlyingBitmap)
      ref.close()
    }
    verify(imageReference).close()
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose_whenCloseableStaticBitmapClosed_thenReturnNull() {
    whenever(closeableStaticBitmap.isClosed()).thenReturn(true)
    whenever(closeableStaticBitmap.cloneUnderlyingBitmapReference()).thenReturn(null)

    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(imageReference)

    // We only detach the reference and do not care if the bitmap is valid
    Assertions.assertThat(extractedReference).isNull()
    verify(imageReference).close()
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose_whenImageReferenceInvalid_thenReturnNull() {
    whenever(imageReference.isValid).thenReturn(false)

    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(imageReference)

    // We only detach the reference and do not care if the bitmap is valid
    Assertions.assertThat(extractedReference).isNull()
    verify(imageReference).close()
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose_whenInputNull_thenReturnNull() {
    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(null)

    Assertions.assertThat(extractedReference).isNull()
    verifyNoMoreInteractions(imageReference)
  }

  @Test
  @Throws(Exception::class)
  fun testExtractAndClose_whenCloseableStaticBitmapNull_thenReturnNull() {
    whenever(imageReference.get()).thenReturn(null)

    val extractedReference: CloseableReference<Bitmap>? =
        FrescoFrameCache.convertToBitmapReferenceAndClose(imageReference)

    Assertions.assertThat(extractedReference).isNull()
    verify(imageReference).close()
  }
}
